### DISCUSSION TITLE
Composer can access private package

### DISCUSSION THREAD

**Discussion by basz** [renovatebot/renovate#17096] [Request Help] [answered]
Hi,

I get an error which seems to prevent lock files from being updated.

I use renovate from github app.
it is a PHP project
the repository is private
a referenced private repository can't be checked out.

first question: I read a lot about config.js. Is it true this only applies to self hosted renovate...? Since I am using renovate as a github app, i can skip these documentations?

This issue must be something about tokens.

I've tried to set action secret as COMPOSER_AUTH to `{"gitlab.com": "privatetoken"}` but that does not seem to work.

composer.json contains (one private, 3 public ones)

```
  "repositories": [
    {
      "type": "git",
      "url": "git@github.com:plhw/lab-api-model-valueobject.git"
    },
    {
      "type": "git",
      "url": "git@github.com:bushbaby/console-application.git"
    },
    {
      "type": "git",
      "url": "git@github.com:basz/zfr-oauth2-server.git"
    },
    {
      "type": "git",
      "url": "git@github.com:basz/zfr-oauth2-server-doctrine.git"
    }
  ],
```

Renovate error:

```
Command failed: docker run --rm --name=renovate_php --label=renovate_child -v "/mnt/renovate/gh/plhw/lab-api":"/mnt/renovate/gh/plhw/lab-api" -v "/tmp/renovate-cache":"/tmp/renovate-cache" -e COMPOSER_CACHE_DIR -e COMPOSER_AUTH -e BUILDPACK_CACHE_DIR -w "/mnt/renovate/gh/plhw/lab-api" docker.io/renovate/php:7.4.30 bash -l -c "install-tool composer 2.3.10 && composer update ibericode/vat --with-dependencies --ignore-platform-req='ext-*' --ignore-platform-req='lib-*' --no-ansi --no-interaction --no-scripts --no-autoloader --no-plugins"
Loading composer repositories with package information

In Git.php line 484:
                                                                               
  Failed to execute git clone --mirror -- 'https://ghs...agq:x-oauth-basic@gi  
  thub.com/plhw/lab-api-model-valueobject.git' '/tmp/renovate-cache/others/co  
  mposer/vcs/git-github.com-plhw-lab-api-model-valueobject.git/'               
                                                                               
  Cloning into bare repository '/tmp/renovate-cache/others/composer/vcs/git-g  
  ithub.com-plhw-lab-api-model-valueobject.git'...                             
  remote: Support for password authentication was removed on August 13, 2021.  
  remote: Please see https://docs.github.com/en/get-started/getting-started-w  
  ith-git/about-remote-repositories#cloning-with-https-urls for information o  
  n currently recommended modes of authentication.                             
  fatal: Authentication failed for 'https://github.com/plhw/lab-api-model-val  
  ueobject.git/'                                                               
                                                                               

update [--with WITH] [--prefer-source] [--prefer-dist] [--prefer-install PREFER-INSTALL] [--dry-run] [--dev] [--no-dev] [--lock] [--no-install] [--no-autoloader] [--no-suggest] [--no-progress] [-w|--with-dependencies] [-W|--with-all-dependencies] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--apcu-autoloader-prefix APCU-AUTOLOADER-PREFIX] [--ignore-platform-req IGNORE-PLATFORM-REQ] [--ignore-platform-reqs] [--prefer-stable] [--prefer-lowest] [-i|--interactive] [--root-reqs] [--] [<packages>...]
```

**Comment by Gabriel-Ladzaretti** [comment]
> first question: I read a lot about config.js. Is it true this only applies to self hosted renovate...? Since I am using renovate as a github app, i can skip these documentations?

That's true, `config.js` is only applicable when self hosted. when using the app it is maintained internally by the app itself.

**Reply by rarkins** [reply]
And Actions secrets only apply to GitHub Actions. Renovate runs entirely separately and does not access them

**Reply by basz** [reply]
Does renovate run as 'me' then? How does it have access to resources that are private?

**Comment by rarkins** [comment]
Is the private repo in the same org as the other repo, or in a different one?

**Reply by basz** [reply]
same

**Reply by rarkins** [reply]
If both repos are private, and you have installed Renovate into both, I would expect it to work automatically. Are they both private and installed?

**Reply by basz** [reply]
yes, i found that hint too. on both repos renovate is renovating...

**Comment by basz** [comment]
from app.renovatebot.com

seems renovate tries to get from packagist ignoring the repositories key in composer.json

```
DEBUG: Found package lookup warnings
{
  "warnings": [
    "Failed to look up dependency bushbaby/console-application",
    "Failed to look up dependency plhw/hf-model-valueobject",
    "Failed to look up dependency ghcr.io/plhw/lab-docker/composer",
    "Failed to look up dependency ghcr.io/plhw/lab-docker/php"
  ],
  "warningFiles": [
    "composer.json",
    "docker-compose.yml"
  ]
}
```

**Reply by basz** [reply]
I've moved github.com:bushbaby/console-application.git to git@github.com:plhw/console-application.git so it's the same organisation (and renovate is active on it)

**Reply by rarkins** [reply]
If Renovate isn't looking up the correct registry then that implies that it's probably not a private registry problem and instead a "knowing where to look" problem. Can you reproduce it using two public repos on github.com?

**Reply by basz** [reply]
Well why should renovate not be able to find these composer repositories as production installations work fine for me.

If I read the [following](https://getcomposer.org/doc/05-repositories.md#using-private-repositories) i see a remark about private repositories 'The only requirement is the installation of SSH keys for a git client'. How does renovate get this when I cannot set  secrets to for the runner?

**Reply by rarkins** [reply]
If you require SSH keys for authentication then it's not compatible with the hosted model, because apps are not given SSH keys/credentials. Apps are given tokens which work over HTTPS (including git over HTTPS)

**Comment by basz** [comment, marked as answer]
Hmm, after some research i found the following two changes enabled renovate to download the repositories and thus create a composer.lock file in resulting PR's.

1. Change `"type": "git",` to ` "type": "vcs",` which is a catch all 'versioning control systems'. Composer the figures out we use git. Appearently renovatebot need's that.

2. Add a name property specifying the exact name of the package `"name": "plhw/hf-model-valueobject",`. The actual location of the repository is `plhw/lab-api-model-valueobject.git`.

I still see issues for these packages in the logs of renovateapp, however the process completes correctly and an updated lock file is created.

```
DEBUG: Datasource unknown error
{
  "datasource": "git-tags",
  "packageName": "git@github.com:plhw/lab-api-model-valueobject.git",
  "err": {
    "task": {
      "commands": [
        "ls-remote",
        "git@github.com:plhw/lab-api-model-valueobject.git"
      ],
      "format": "utf-8",
      "parser": "[function]"
    },
    "message": "ERROR: Repository not found.\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n",
    "stack": "Error: ERROR: Repository not found.\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n\n    at Object.action (/home/ubuntu/renovateapp/node_modules/simple-git/dist/cjs/index.js:1194:25)\n    at PluginStore.exec (/home/ubuntu/renovateapp/node_modules/simple-git/dist/cjs/index.js:1229:29)\n    at /home/ubuntu/renovateapp/node_modules/simple-git/dist/cjs/index.js:1592:43\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/home/ubuntu/renovateapp/node_modules/simple-git/dist/cjs/index.js:1590:16)\n    at GitExecutorChain.<anonymous> (/home/ubuntu/renovateapp/node_modules/simple-git/dist/cjs/index.js:1574:44)\n    at Generator.next (<anonymous>)\n    at fulfilled (/home/ubuntu/renovateapp/node_modules/simple-git/dist/cjs/index.js:55:24)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)"
  }
}
DEBUG: Failed to look up dependency plhw/hf-model-valueobject (plhw/hf-model-valueobject)
```
